### PR TITLE
added algorithm 1 and tracked diffs between papers

### PR DESCRIPTION
**Pull request by bdavidzhang**
Summary

  Implements D4RT's Algorithm 1 (occupancy-grid efficient dense all-pixel tracking) as an inference-time scheduler over
  the frozen model, plus a CLI to run it.

  Motivation / Context

  The paper's headline efficiency trick existed only as a config string (occupancy_grid_tracking_all_pixels in
  configs/model_effective.yaml:74-76) with no implementation. It's a pure inference-time scheduler — it exploits
  spatio-temporal redundancy by only seeding tracks from pixels not yet covered by an existing track — so it needs no 
  retraining or weight changes. This wires up the dense-reconstruction path the repo was missing.

  Changes Made

  - infer_track_3d.py: add dense_track_occupancy_grid(...) — encodes the clip once (F = E(V)), then iteratively seeds
  un-visited grid cells, decodes full per-frame tracks, and marks visibly-covered cells visited (paper Alg. 1, lines
  1–13). Reuses the existing _encode_model_memory / _run_model_for_queries / _build_query_for_targets query path.
  - infer_track_3d.py: add _decode_tracks_for_sources(...) helper.
  - infer_dense_track.py: new CLI — loads a checkpoint, takes a WorldTrack .npz or video file, writes *_dense_tracks.npz
  + diagnostics JSON, prints the speedup.
  - docs/paper_vs_code_gap.md: paper-vs-code gap analysis; marks Alg. 1 as now implemented.

  How to Test

  EXP=checkpoints/OpenD4RT_32CLIP_9Dataset_NoAUG
  python infer_dense_track.py \
    --model-config "$EXP/model.yaml" --ckpt-path "$EXP/opend4rt.ckpt" \
    --input data/worldtrack_release/pstudio_mini/softball_25.npz \
    --grid 48 --seed-batch 512 --mark-radius 1 --output-dir tmp/dense_track
  Logic-only smoke test (mock model, no checkpoint needed) verified: correct shapes, clean termination, clip-length
  guard, and the expected adaptive speedup.

  Results
  
  Mock-model smoke test (grid 8×8, 4 frames): 64 tracks vs 256 naive cells → 4× seed + decoder-query speedup,
  occupancy_filled_ratio = 1.0. Real-checkpoint speedup is content-dependent (paper reports 5–15×); output is the same
  per-track quality as naive querying.
  
  Notes / Limitations

  - Within-clip only: processes one encoded clip (raises if num_frames > clip_frames). Long videos need composing with
  the existing anchor-clip/Sim3 stitching — not wired here.
  - It's the redundancy approximation by design: pixels covered by a neighbor's track don't get an independent trajectory
  (that's the source of the speedup).
  - 3D tracks are returned in reference_frame's camera (unified frame for point clouds); occupancy marking uses the uv_2d
  + visibility heads.
  - Not yet validated against the 14 GB checkpoint (no GPU/download in this session); the model-call path reuses
  already-tested helpers.

  Related Work / References
  
  - Paper: Efficiently Reconstructing Dynamic Scenes One D4RT at a Time — Algorithm 1 (docs/D4RT_paper.pdf, p.3)
  - Checkpoints: Lijiaxin0111/OpenD4RT (https://huggingface.co/Lijiaxin0111/OpenD4RT)
  - See docs/paper_vs_code_gap.md

### CODE REVIEW

File: infer_dense_track.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+from __future__ import annotations
+
+import argparse
+import json
+from pathlib import Path
+
+import numpy as np
+import torch
+
```
[the remark this answers is not in the document]
Comment (bdavidzhang) [reply infer_dense_track.py:27]: this is fair
